miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/MiniAppDownloader.kt

```diff
@@ -148,15 +148,24 @@ internal class MiniAppDownloader(
     @VisibleForTesting
     fun prepareMiniAppManifest(metadataEntity: MetadataEntity): MiniAppManifest {
         val requiredPermissions = metadataEntity.metadata?.requiredPermissions?.map {
-            Pair(MiniAppCustomPermissionType.getValue(it.name), it.reason)
+            Pair(MiniAppCustomPermissionType.getValue(it.name ?: ""), it.reason ?: "")
         } ?: emptyList()
 
         val optionalPermissions = metadataEntity.metadata?.optionalPermissions?.map {
-            Pair(MiniAppCustomPermissionType.getValue(it.name), it.reason)
+            Pair(MiniAppCustomPermissionType.getValue(it.name ?: ""), it.reason ?: "")
         } ?: emptyList()
 
+        val requiredFilteredValues = requiredPermissions.toMutableList()
+        requiredFilteredValues.removeAll { (first) ->
+            first.type == MiniAppCustomPermissionType.UNKNOWN.type
+        }
+        val optionalFilteredValues = optionalPermissions.toMutableList()
+        optionalFilteredValues.removeAll { (first) ->
+            first.type == MiniAppCustomPermissionType.UNKNOWN.type
+        }
+
```

Comment:
I think you can do a check `it.name` null or not before adding item to a list -> we don't need to do filter the values.

Reply:
I think, we can do the null check using `?:` and I am filtering the unknown values to eliminate as those are unnecessary.

Reply:
I have created the [PR](https://github.com/khairul-alam-licon/android-miniapp/pull/5) for this. Hope it help.

Reply:
Thank you for the PR to understand the point of views:
Here are my feedback [on the PR](https://github.com/khairul-alam-licon/android-miniapp/pull/5/files)

**SDK: MiniAppDownloader.prepareMiniAppManifest**

- `map` & `forEach` usually iterates over all the items, the significant difference is `map` returns a list containing the results. On the other hand, `forEach` requires to performs the operation on each item by walking inside the loop. I was using `map` twice & your PR is also using `forEach` twice by calling `getList`, no? It means, the operation are still the same. Moreover, `map` is chainable e.g. we can attach other operations like `flatMap`, `filter`, `groupBy` ... , so it's good to use if possible.
- Because of using `getList` now we have two functions, but we can make a single function using `map` like what I am already doing. 

**Sample app: PreloadMiniAppWindow**
- Thank you for pointing that we don't need `SharedPreferences` to store acceptance. I want to adopt this change.

**Sample app: PreloadMiniAppViewModel**
- I don't know what does it mean by `_requiredShowManifest`, we don't have any concept like that, rather we have "required permission" concept. 
- The placements of some of the functions into ViewModel is okay, maybe I want to discuss about the architecture pattern in a separate session, but I want to mention that in future, we will have more UI improvement, then we will need to refactor e.g. in some usecases, we'll may need to trigger something to launch the screen from the viewmodel according to the changes from your PR. 

**Another discussion for clarity**, [not related to the PR](https://github.com/khairul-alam-licon/android-miniapp/pull/5/files).
- This feature will not work without valid `versionId`, it means we don't need to initiate this screen from `MiniAppInputActivity` anymore, rather we can now move it to `MiniAppDisplayActivity` because the flow has been changed inside `MiniApp.create`. After moving it to `MiniAppDisplayActivity`, we can finally utilize `RequiredPermissionsNotGrantedException` and then we will no need to use `isAcceptedRequiredPermissions` function what I introduced. 

After considering those things, I feel that we can make a separate PR later to address all of these key issues. I want to avoid this time since we are having a tight schedule. 

@minh-rakuten @corycaywood What do you think?

Reply:
1/ SDK: MiniAppDownloader.prepareMiniAppManifest

Does the refactor do the same as this current implementation? Which one gives better time and space complexity? Two functions here but it's shorter and gives better performance.

2/ PreloadMiniAppWindow & PreloadMiniAppViewModel
I'm trying to correct the function of viewmodel and view here. `_requiredShowManifest` is the purpose of this screen. You mentioned that we show this dialog again when there are changes in manifest, not only required permissions.

Reply:
I think that `_miniAppManifest` is enough to express the purpose instead of using `_requiredShowManifest`. But as long as it doesn't break the functionality its ok

Reply:
I think that `_miniAppManifest` is enough to express the purpose instead of using `_requiredShowManifest`. But as long as it doesn't break the functionality its okay.

Reply:
@minh-rakuten Thanks for the discussions and affords, I have merged your changes, I have validated it, it works well 👍